src/proofs/inner_product.rs

```diff
@@ -218,6 +236,99 @@ impl InnerProductArg {
             Err(InnerProductError)
         }
     }
+
+    ///
+    /// Returns Ok() if the given inner product satisfies the verification equations,
+    /// else returns `InnerProductError`.
+    /// 
+    /// Uses a single multiexponentiation (multiscalar multiplication in additive notation)
+    /// check to verify an inner product proof. 
+    /// 
+    pub fn fast_verify(&self, g_vec: &[GE], hi_tag: &[GE], ux: &GE, P: &GE)
+    -> Result<(), Errors> {
+        let G = &g_vec[..];
+        let H = &hi_tag[..];
+        let n = G.len();
+        let order = FE::q();
+
+        // All of the input vectors must have the same length.
+        assert_eq!(G.len(), n);
+        assert_eq!(H.len(), n);
+        assert!(n.is_power_of_two());
+
+        let lg_n = self.L.len();
+
+        let mut x_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut x_inv_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut minus_x_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut minus_x_inv_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut allinv = BigInt::one();
+        for (Li, Ri) in self.L.iter().zip(self.R.iter()) {
+
+            let x = HSha256::create_hash_from_ge(&[&Li, &Ri, &ux]);
+            let x_bn = x.to_big_int();
+            let x_inv_fe = x.invert();
+            let x_inv_bn = x_inv_fe.to_big_int();
+            let x_sq_bn = BigInt::mod_mul(&x_bn, &x_bn, &order);
+            let x_inv_sq_bn =
+                BigInt::mod_mul(&x_inv_fe.to_big_int(), &x_inv_fe.to_big_int(), &order);
+            // let x_sq_fe: FE = ECScalar::from(&x_sq_bn);
+            // let x_inv_sq_fe: FE = ECScalar::from(&x_inv_sq_bn);
+
+            x_sq_vec.push(x_sq_bn.clone());
+            x_inv_sq_vec.push(x_inv_sq_bn.clone());
+            minus_x_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_sq_bn, &order));
+            minus_x_inv_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_inv_sq_bn, &order));
+            allinv = allinv * x_inv_bn;
+        }
+
+        let mut s: Vec<BigInt> = Vec::with_capacity(n);
+        s.push(allinv);
+        for i in 1..n {
+            let lg_i = (32 - 1 - (i as u32).leading_zeros()) as usize;
```

Comment:
what is the meaning of the number 32 ?

Reply:
As explained in my answer to your next question, once we compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_{1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_{1}" title="s_{1}" /></a>, we need to compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_{2},&space;s_3,&space;\dots" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_{2},&space;s_3,&space;\dots" title="s_{2}, s_3, \dots" /></a> according to the bits in <a href="https://www.codecogs.com/eqnedit.php?latex=(i-1)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(i-1)" title="(i-1)" /></a> for <a href="https://www.codecogs.com/eqnedit.php?latex=i&space;=&space;2,3,\dots" target="_blank"><img src="https://latex.codecogs.com/gif.latex?i&space;=&space;2,3,\dots" title="i = 2,3,\dots" /></a>.

Here, I am iterating (i-1) as 1,2,...(n-1). 32 just means that I am assuming that 

<a href="https://www.codecogs.com/eqnedit.php?latex=\text{log}_2n&space;\le&space;32&space;\implies&space;n&space;\le&space;2^{32}&space;\approx&space;4.29&space;\text{&space;Billion}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\text{log}_2n&space;\le&space;32&space;\implies&space;n&space;\le&space;2^{32}&space;\approx&space;4.29&space;\text{&space;Billion}" title="\text{log}_2n \le 32 \implies n \le 2^{32} \approx 4.29 \text{ Billion}" /></a>

I should have mentioned it in a comment in the code. Hope this helps.

Reply:
ok, so why you assume 32 and not take as a parameter ? 
- I think you should add a check for n and take the next number as this parameter
- or , if you are convinced that it must be 32 you should add a bound check, throwing error in case n is too large and exit.

Reply:
You are right. We don't need 32 necessarily. We can take it as a parameter. I think it would be a bit confusing for a user if we let the user decide this. Instead I can make it 64 (which is far more than enough according to me) and add a check and throw an error if sizes greater than 64 are encountered.

Reply:
can't it be a function of `n` ?

Reply:
Oh right. I am just using it in computing log_2(i). So it should work even when it depends on n. I will try setting it just to log_2(n).

Reply:
So a constant like 32 is necessary because I am using it to compute floor of log_2 of integers 1,2,..,(n-1). The way I am doing this is as follows:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;&&space;i=13&space;\text{&space;as&space;u32&space;}&space;\implies&space;\&space;\overbrace{&space;000\dots&space;00}^{28}1101&space;\\&space;&&space;\qquad&space;\lfloor&space;\text{log}_2(i)&space;\rfloor&space;=&space;32-1-28&space;=&space;3&space;\end{align*}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;&&space;i=13&space;\text{&space;as&space;u32&space;}&space;\implies&space;\&space;\overbrace{&space;000\dots&space;00}^{28}1101&space;\\&space;&&space;\qquad&space;\lfloor&space;\text{log}_2(i)&space;\rfloor&space;=&space;32-1-28&space;=&space;3&space;\end{align*}" title="\begin{align*} & i=13 \text{ as u32 } \implies \ \overbrace{ 000\dots 00}^{28}1101 \\ & \qquad \lfloor \text{log}_2(i) \rfloor = 32-1-28 = 3 \end{align*}" /></a>

Now we could also use 64 instead. But it necessarily has to be 32 (because of u32) or 64 (for u64). I think I will just add a check for <a href="https://www.codecogs.com/eqnedit.php?latex=n&space;>&space;2^{64}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?n&space;>&space;2^{64}" title="n > 2^{64}" /></a>.

Reply:
I have changed it to support 64. Please check.

Reply:
What about using BigInt for it? 
there's a function to get the bit length of a BigInt.

Reply:
The length `n` we compute in verification has the datatype `usize`.
So anyway we cannot have `n > 2^{64}` (for 64-bit targets). Are you suggesting that I try setting `n` itself to BigInt and change the scalar computations accordingly?

Reply:
you make a good point. lets keep the 64bit and not overkill. I will merge.